src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -83,7 +83,12 @@ SotaUptaneClient::SotaUptaneClient(Config &config_in, std::shared_ptr<INvStorage
   }
 }
 
-SotaUptaneClient::~SotaUptaneClient() { conn.disconnect(); }
+SotaUptaneClient::~SotaUptaneClient() {
+    conn.disconnect();
```

Comment:
If this is triggered by SotaUptaneClient's destructor, it will happen even when using the api manually.

During our discussion, the consensus was to only trigger it in Aktualizr's "automatic" mode, i.e. from `Aktualizr::RunForever` but it was probably not made very clear. @patrickvacek is that what you understood?

But I see that doing it from this destructor is convenient since it happens after flushing the report queue. Maybe a bit ugly but this code could be moved after `while (!shutdown_) {}` in RunForever and somehow do a manual flush before that?

Reply:
Yes, I agree about only doing this in the "automatic" mode, that was also my understanding.

I'd originally thought that we would trigger the reboot in a `postInstall` function in the package managers, but since that has the problem of flushing the queue, perhaps doing it in the Aktualizr class is better. (And I apologize @mike-sul that that is closer to where you originally had it!)

Reply:
Yes, the original consensus was to trigger it at the end of UptaneCycle within Aktualizr::RunForever. 
But, we also realized/agreed that it won't solve the sync issue, i.e. rebooting a device while two other active threads (the API and Report threads) might do some actions interruption of which can cause a problem during the following aktualizr run/start.  e.g. rebooting during writing to the DB or some report event wasn't sent to the backend so the local DB and the backend's DN are not in sync. I am speculating here of course since I haven't learned the code & design very well yet, so you should know better if my statement is correct.
My thought process around the sync issue resolution was the following:

1) Graceful reboot (no any thread & action interruption) can be done only after graceful exit from the three active/looping threads running withing the aktualizr process;

2) the UptainCycle thread can be exited gracefully by setting the shutdown flag hence I added Shutdown call;

3) the API thread is shutdown-ed by the aforementioned Aktualizr::Shutdown() call but the thread is "joined" at the Aktualizr dtor so we cannot reboot before (if we wanna clean/graceful reboot)

4) the Report thread is shutdown-ed and "joined" at the ReportQueue object dtor lifetime of which is owned by SotaUptaneClient object lifetime of which in turn is owned by  Aktualizr object created on a stack of the main thread and being destroyed only when the main thread goes out of the main block.

Therefore, within the current implementation the correct place/moment for the clean/graceful reboot call is at the ReportQueue dtor.

If there is no need in the clean/graceful reboot in this case then we can trigger the reboot from whatever place. If it's important to have the graceful reboot then it should be done at the ReportQueue dtor or somewhere down of a control flow. Also, there is an option to refactor the code slightly and trigger the graceful reboot in more appropriate place.

But, again, I am a new guy here and haven't learned the code very well, so I can be incorrect in my statements :).